pkg/cloudmap/resource_manager.go

```diff
@@ -0,0 +1,444 @@
+	Cleanup(ctx context.Context, vn *appmesh.VirtualNode) error
+}
+
+func NewCloudMapResourceManager(
+	k8sClient client.Client,
+	cloudMapSDK services.CloudMap,
+	virtualNodeEndpointResolver VirtualNodeEndpointResolver,
+	cloudMapInstanceReconciler CloudMapInstanceReconciler,
+	accountID string,
+	log logr.Logger) ResourceManager {
+
+	return &cloudMapResourceManager{
+		k8sClient:                   k8sClient,
+		cloudMapSDK:                 cloudMapSDK,
+		virtualNodeEndpointResolver: virtualNodeEndpointResolver,
+		cloudMapInstanceReconciler:  cloudMapInstanceReconciler,
+		namespaceIDCache:            cache.NewLRUExpireCache(cloudMapNamespaceCacheMaxSize),
+		serviceIDCache:              cache.NewLRUExpireCache(cloudMapServiceCacheMaxSize),
+		accountID:                   accountID,
+		log:                         log,
+	}
+}
+
+// cloudMapResourceManager implements ResourceManager
+type cloudMapResourceManager struct {
+	k8sClient                   client.Client
+	cloudMapSDK                 services.CloudMap
+	virtualNodeEndpointResolver VirtualNodeEndpointResolver
+	cloudMapInstanceReconciler  CloudMapInstanceReconciler
+	namespaceIDCache            *cache.LRUExpireCache
+	serviceIDCache              *cache.LRUExpireCache
+	accountID                   string
+	log                         logr.Logger
+}
+
+type cloudMapNamespaceSummary struct {
+	NamespaceID   string
+	NamespaceType string
+}
+
+type cloudMapServiceSummary struct {
+	NamespaceID string
+	ServiceID   string
+}
+
+const (
+	cloudMapNamespaceCacheMaxSize = 100
+	cloudMapNamespaceCacheTTL     = 2 * time.Minute
+	cloudMapServiceCacheMaxSize   = 1024
+	cloudMapServiceCacheTTL       = 2 * time.Minute
+)
+
+func (m *cloudMapResourceManager) Reconcile(ctx context.Context, vNode *appmesh.VirtualNode) error {
+
+	cloudMapConfig := vNode.Spec.ServiceDiscovery.AWSCloudMap
+	creatorRequestID := vNode.UID
+
+	serviceSummary, err := m.getOrCreateCloudMapService(ctx, cloudMapConfig, string(creatorRequestID))
+	if err != nil {
+		m.log.Error(err, "CloudMap RM: failed to create cloudMap Service for", "vNode: ", vNode.Name)
+		return err
+	}
+
+	readyPods, notReadyPods, ignoredPods, err := m.virtualNodeEndpointResolver.ResolveCloudMapEndPoints(ctx, vNode)
+	if err != nil {
+		m.log.Error(err, "CloudMap RM: failed to get pods for", "vNode: ", vNode.Name)
+		return err
+	}
+
+	//Reconcile pod instances with Cloudmap
+	if err := m.cloudMapInstanceReconciler.ReconcileCloudMapInstances(ctx, readyPods, notReadyPods, ignoredPods,
+		serviceSummary.ServiceID, serviceSummary.NamespaceID, vNode); err != nil {
+		log.Error(err, "CloudMap RM: Error reconciling instances with CloudMap")
+		return err
+	}
+	return nil
+}
+
+func (m *cloudMapResourceManager) Cleanup(ctx context.Context, vNode *appmesh.VirtualNode) error {
+	cloudMapConfig := vNode.Spec.ServiceDiscovery.AWSCloudMap
+	if err := m.deleteCloudMapService(ctx, vNode, cloudMapConfig); err != nil {
```

Comment:
seem we didn't call ReconcileCloudMapInstances here. (which we should?)

Reply:
yeah it is called from the delete function